DIPs/DIP1005.md

```diff
@@ -0,0 +1,781 @@
+
+```d
+    static import std.range.primitives;
+    alias isInputRange = std.range.primitives.isInputRange;
+    alias isForwardRange = std.range.primitives.isInputRange;
+    ...
+```
+
+    Such scaffolding is of course undesirable in the first place. Also, at least by the current language rules, such `alias` definitions would need to load the module anyway so as to ensure the name does exist. In order for this idiom to work, it would require subtle changes to the language that specify how certain `alias` declarations are exempt from early checking and delayed to the first actual use.
+* In either setup, `import`s are collapsed into their union, usually at the top of the module. Such lists grow out of sync with the actual code because during maintenance the programmer working on one declaration is not motivated to simultaneously alter a module-level `import` list shared by all declarations in the module. Over time, the `import`s grow into a superset of the actual depedencies used by the code, and do not reflect which declarations cause which imports even when accurate.
+* The "carrying" aspect is lost: any migration of a declaration to another module must be followed by awkwardly doing surgery on the `import` list of the receiving module. Again, the migration may leave unused `import`s in the module the declaration is taken from. The only recourse to keeping the `import` list in sync is special tooling or time-consuming discipline (search the module for uses, attempt recompilation).
+
+Although we consider introducing lazy imports an improvement over the current state of affairs, our assessment is that such a feature would fall short of truly allowing a project to rein in its dependency structure.
+
+We have experimented with converting the standard library module `std.array` to one of the two idioms. Conversion to either the "static import" form or the "selective import" form may be achieved by brute force through dedicated tooling: first generate code that enumerates all symbols in the module, then eliminate them one by one and attempt to rebuild. Such an approach is time-consuming and would be only used at rare intervals.
+
+The manual conversion of `std.array` to the "static import" form is shown [here](https://github.com/dlang/phobos/pull/4962). It leads to the expected lengthening of the symbols used in declarations, which appears to eliminate one disadvantage by introducing another. Also the manual conversion process turned out to be prohibitively difficult; we would only recommend this conversion using automated tooling.
+
+The manual conversion of `std.array` to the "selective import" form is shown [here](https://github.com/dlang/phobos/pull/4963). Conversion was successful but because it collapses all `import`s at the top, it does not make it much easier to identify e.g. what dependencies would be pulled if a given artifact in `std.array` were used. Again the manual process was highly nontrivial.
```

Comment:
Is this really missing a conversion to `with (import mod)`?
I'd assume the same arguments apply, plus how would you even test that the "with imports" are complete, an overlooked module level import would hide any errors.